### PR TITLE
tiles: add new store tile version

### PR DESCRIPTION
**Pull request by arjain4**

### CODE REVIEW

File: src/app/fdctl/Local.mk [hunk 1]
```diff
@@ -57,6 +58,7 @@ $(OBJDIR)/obj/app/fdctl/run/tiles/fd_shred.o: src/app/fdctl/run/tiles/generated/
 $(OBJDIR)/obj/app/fdctl/run/tiles/fd_verify.o: src/app/fdctl/run/tiles/generated/verify_seccomp.h
 $(OBJDIR)/obj/app/fdctl/run/tiles/fd_metric.o: src/app/fdctl/run/tiles/generated/metric_seccomp.h
 $(OBJDIR)/obj/app/fdctl/run/tiles/fd_sign.o: src/app/fdctl/run/tiles/generated/sign_seccomp.h
+$(OBJDIR)/obj/app/fdctl/run/tiles/fd_sign.o: src/app/fdctl/run/tiles/generated/store_seccomp.h
```
Comment (mmcgee-jump) [inline src/app/fdctl/Local.mk:61]: should be store_int_seccomp ? and the left side is also bugged

File: src/disco/topo/fd_topo.h [hunk 2]
```diff
@@ -210,6 +210,11 @@ typedef struct {
     struct {
       ulong accounts_cnt;
     } benchg;
+
+    struct {
+      ulong snapshot_slot;
+      char  identity_key_path[ PATH_MAX ];
+    } store;
```
Comment (mmcgee-jump) [inline src/disco/topo/fd_topo.h:217]: store_int ?

File: src/disco/fd_disco_base.h [hunk 3]
```diff
@@ -87,6 +90,18 @@ FD_FN_CONST static inline ulong fd_disco_poh_sig_pkt_type( ulong sig ) { return
 FD_FN_CONST static inline ulong fd_disco_poh_sig_slot( ulong sig ) { return (sig >> 8); }
 FD_FN_CONST static inline ulong fd_disco_poh_sig_bank_tile( ulong sig ) { return (sig >> 2) & 0x3FUL; }
 
+FD_FN_CONST static inline ulong
+fd_disco_replay_sig( ulong slot,
+                     ulong flags ) {
+   /* The high 7 bits of the low byte of the signature field is the bank
```
Comment (mmcgee-jump) [inline src/disco/fd_disco_base.h:96]: Comment seems wrong, and do you need functions to extract the slot and flags as well?

Comment (arjain4) [reply src/disco/fd_disco_base.h:96]: That will get added with the replay tile changes in a follow up

File: src/app/fdctl/Local.mk [hunk 4]
```diff
@@ -57,6 +58,7 @@ $(OBJDIR)/obj/app/fdctl/run/tiles/fd_shred.o: src/app/fdctl/run/tiles/generated/
 $(OBJDIR)/obj/app/fdctl/run/tiles/fd_verify.o: src/app/fdctl/run/tiles/generated/verify_seccomp.h
 $(OBJDIR)/obj/app/fdctl/run/tiles/fd_metric.o: src/app/fdctl/run/tiles/generated/metric_seccomp.h
 $(OBJDIR)/obj/app/fdctl/run/tiles/fd_sign.o: src/app/fdctl/run/tiles/generated/sign_seccomp.h
+$(OBJDIR)/obj/app/fdctl/run/tiles/fd_sign_int.o: src/app/fdctl/run/tiles/generated/store_int_seccomp.h
```
Comment (mmcgee-jump) [inline src/app/fdctl/Local.mk:61]: still needs to be fd_store_int.o on the left

File: src/app/fdctl/run/tiles/fd_store_int.c [hunk 5]
```diff
@@ -0,0 +1,604 @@
...
+  fd_valloc_t valloc = fd_alloc_virtual( alloc );
+  (void)valloc;
+}
+
+static ulong
+populate_allowed_seccomp( void *               scratch,
+                          ulong                out_cnt,
+                          struct sock_filter * out ) {
+  (void)scratch;
+  populate_sock_filter_policy_store( out_cnt, out, (uint)fd_log_private_logfile_fd() );
```
Comment (mmcgee-jump) [inline src/app/fdctl/run/tiles/fd_store_int.c:570]: Does this compile? should be policy_store_int ?

File: src/app/fdctl/run/tiles/fd_store_int.c [hunk 6]
```diff
@@ -0,0 +1,604 @@
...
+
+#define SHRED_IN_IDX    0
+#define REPAIR_IN_IDX   1
+#define STAKE_IN_IDX    2
+#define REPLAY_IN_IDX   3
+
+#define REPAIR_OUT_IDX  0
+#define REPLAY_OUT_IDX  1
+
+#define MAX_REPAIR_PEERS (40200UL)
```
Comment (mmcgee-jump) [inline src/app/fdctl/run/tiles/fd_store_int.c:38]: Are these consensus related? Where do these constants come from?

File: src/app/fdctl/run/tiles/fd_store_int.c [hunk 7]
```diff
@@ -0,0 +1,604 @@
...
+  fd_wksp_t * wksp;
+  fd_wksp_t * blockstore_wksp;
+
+  fd_pubkey_t          identity_key[1]; /* Just the public key */
+
+  fd_store_t * store;
+  fd_blockstore_t * blockstore;
+
+  fd_wksp_t *     net_in;
+  ulong           chunk;
```
Comment (mmcgee-jump) [inline src/app/fdctl/run/tiles/fd_store_int.c:54]:
```suggestion
  ulong           net_in_chunk;
```

File: src/app/fdctl/run/tiles/fd_store_int.c [hunk 8]
```diff
@@ -0,0 +1,604 @@
...
+
+    memcpy( ctx->shred_buffer, shred, sz );
+    *opt_filter = 0;
+    
+    return;
+  }
+
+  *opt_filter = 1;
+
+  return;
```
Comment (mmcgee-jump) [inline src/app/fdctl/run/tiles/fd_store_int.c:199]: Not needed

File: src/app/fdctl/run/tiles/fd_store_int.c (hunk 6 above)
Comment (arjain4) [reply src/app/fdctl/run/tiles/fd_store_int.c:38]: Not consensus related

**Comment by arjain4** [comment]
@mmcgee-jump sorry this was supposed to be a draft. Thank you for the review will clean up

File: src/app/fdctl/run/tiles/fd_store_int.c [hunk 9]
```diff
@@ -0,0 +1,604 @@
...
+  /* Create scratch region */
+  FD_TEST( (!!smem) & (!!fmem) );
+  fd_scratch_attach( smem, fmem, SCRATCH_SMAX, SCRATCH_SDEPTH );
+
+  
+  ctx->wksp = topo->workspaces[ topo->objs[ tile->tile_obj_id ].wksp_id ].wksp;
+  ctx->blockstore_wksp = NULL;
+  for( ulong i = 0; i < topo->wksp_cnt; i++ ) {
+    fd_topo_wksp_t * topo_wksp = &topo->workspaces[i];
+    if( strcmp( topo_wksp->name, "bstore" ) == 0 ) {
```
Comment (mmcgee-jump) [inline src/app/fdctl/run/tiles/fd_store_int.c:448]: I don't think this is quite right. The blockstore needs to be created during `workspace.c` initialization, otherwise a timing race at tile boot could have other tiles not find this properly? If other tiles don't need to find it, why is it in a random workspace?

File: src/app/fdctl/run/tiles/fd_store_int.c [hunk 10]
```diff
@@ -0,0 +1,604 @@
...
+  for( ulong i = fd_pending_slots_iter_init( ctx->store->pending_slots );
+         (i = fd_pending_slots_iter_next( ctx->store->pending_slots, ctx->store->now, i )) != ULONG_MAX; ) {
+    uchar const * block = NULL;
+    ulong         block_sz = 0;
+    ulong repair_slot = FD_SLOT_NULL;
+    int store_slot_prepare_mode = fd_store_slot_prepare( ctx->store, i, &repair_slot, &block, &block_sz );
+
+    ulong slot = repair_slot == 0 ? i : repair_slot;
+    FD_LOG_DEBUG(( "store slot - mode: %d, slot: %lu, repair_slot: %lu", store_slot_prepare_mode, i, repair_slot ));
+    fd_store_tile_slot_prepare( ctx, store_slot_prepare_mode, slot );
```
Comment (mmcgee-jump) [inline src/app/fdctl/run/tiles/fd_store_int.c:401]: Pretty suspicious that this call belongs in housekeeping? I suspect you want this to be in after_credit and have your own timer to fire it periodically.

File: src/app/fdctl/run/tiles/fd_store_int.c [hunk 11]
```diff
@@ -0,0 +1,604 @@
...
+      fd_block_info_t block_info;
+      fd_runtime_block_prepare( block_data, block->data_sz, fd_scratch_virtual(), &block_info );
+    
+      FD_LOG_NOTICE(( "block prepared - slot: %lu", slot ));
+
+      fd_txn_p_t * txns = fd_type_pun( out_buf );
+      ulong txn_cnt = fd_runtime_block_collect_txns( &block_info, txns );
+
+      ulong out_sz = sizeof(ulong) + sizeof(fd_hash_t) + ( txn_cnt * sizeof(fd_txn_p_t) );
+      fd_mcache_publish( ctx->replay_out_mcache, ctx->replay_out_depth, ctx->replay_out_seq, replay_sig, ctx->replay_out_chunk, txn_cnt, 0UL, tsorig, tspub );
```
Comment (mmcgee-jump) [inline src/app/fdctl/run/tiles/fd_store_int.c:364]: Why aren't you using the primary out for these? It seems like they need to be flow controlled and reliable.

**Review by mmcgee-jump — CHANGES_REQUESTED** [review]

File: src/app/fdctl/run/tiles/fd_store_int.c [hunk 12]
```diff
@@ -0,0 +1,604 @@
...
+  ctx->store->now = fd_log_wallclock();
+
+  if( FD_UNLIKELY( in_idx==STAKE_IN_IDX ) ) {
+    fd_stake_ci_stake_msg_fini( ctx->stake_ci );
+    return;
+  }
+
+  if( FD_UNLIKELY( in_idx==SHRED_IN_IDX ) ) {
+    for( ulong i = 0; i < ctx->s34_buffer->shred_cnt; i++ ) {
+      if( fd_store_shred_insert( ctx->store, &ctx->s34_buffer->pkts[i].shred ) < FD_BLOCKSTORE_OK ) {
```
[the remark this answers is not in the document]
Comment (arjain4) [reply src/app/fdctl/run/tiles/fd_store_int.c:219]: The return value of the library api is setup in a strange way here. Will add a todo to look into this

File: src/app/fdctl/run/tiles/fd_store_int.c (hunk 9 above)
Comment (arjain4) [reply src/app/fdctl/run/tiles/fd_store_int.c:443]: The blockstore is shared across tiles. This will be clearer when the replay tile is added

Comment (lheeger-jump) [reply src/app/fdctl/run/tiles/fd_store_int.c:408]: There is not a timing race. join is safe.

Comment (mmcgee-jump) [reply src/app/fdctl/run/tiles/fd_store_int.c:408]: Join is not safe if it's called before new, who calls new? It's impossible for me to review this stuff in isolation, but this really doesn't look right.
